### PR TITLE
[WIP] Time series modules 

### PR DESCRIPTION
**Pull request by hesenp**
Some time series wrapper as discussed in #1361 . Could appreciate any discussions and inputs as the effort is very nascent.

A few things this model tries to accomplish: 
 1. provide a building blocks for time series analysis. i think this should not be limited to forecasting only. it should have interpretation capability as well. 
 2. the building blocks should fit into the `fit`, `predict` pattern like sci-kit learn. 
 3. hide away sampling procedures. 
 4. hide away some of the numpyro handler functions, which could be a bit daunting to new users (including myself) 
 5. still enable flexibility to try out different patterns. 

With the above objectives in mind, this PR (so far) includes an implementation of a naive AR(1) model that is built on the following procedures: 

 * method 1: low level implementation of model, based on `NUTModelTrainer` only. 
 * method 2: higher level implementation of model, based on `StateSpaceModel`. 

Could appreciate any inputs and hole poking on whether these patterns in method 1 and 2 would break in certain situations. I might come back and add more text book examples later.

### CODE REVIEW

File: numpyro/contrib/time_series.py [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+        return self.mcmc
+
+    def predict(self, data, num_samples=100): 
+        posterior = Predictive(self.model, num_samples=num_samples)
+        posterior_sample = posterior(self.rng_key)
+        return posterior_sample
+
+
+
+class StateSpaceModel(NUTModelTrainer):
```
Comment (hesenp) [inline numpyro/contrib/time_series.py:40]: btw, is it correct to call this pattern state space model?

**Review by fehiepsi — COMMENTED** [review]
Thanks, @hesenp! For low-level implementation, we should aim for flexibility while for high-level API, I prefer easy-to-use. The current designs look reasonable to me. I just wish that we can support both SVI and NUTS forecasters (personally, I would like to use SVI to quickly fit the model first). Maybe having a method `fit_svi` or `fit(..., guide=None)` to run svi?

**Comment by fehiepsi** [comment]
Maybe it is fine to not support `SVI` in time-series module. We can have a separate `forecast` module that follows Pyro semantic: Forecaster and HMCForecaster.

**Comment by hesenp** [comment]
ah sorry i'm very slow these days. will get back to it ha.

**Comment by fehiepsi** [comment]
Closed because it is better (for api design and maintanance) to have a separate library for this. See also https://github.com/pyro-ppl/numpyro/issues/1361

**Pull request closed without merging**